diff --git a/include/verilated_types.h b/include/verilated_types.h
index 40847b65e..13909a5d6 100644
--- a/include/verilated_types.h
+++ b/include/verilated_types.h
@@ -2070,7 +2070,10 @@ struct VlNull final {
         return nullptr;
     }
 };
+// Clang tends to throw ambiguous operator '==' error, even when other candidates require a cast
+#if !defined(__clang__)
 inline bool operator==(const void* ptr, VlNull) { return !ptr; }
+#endif

 //===================================================================
 // Verilog class reference container
