diff --git a/riscv/BUILD b/riscv/BUILD
index 0189fd9..e530a3b 100644
--- a/riscv/BUILD
+++ b/riscv/BUILD
@@ -136,11 +136,9 @@ cc_library(
     copts = select({
         "darwin_arm64_cpu": [
             "-O3",
-            "-ffp-model=strict",
         ],
         "//conditions:default": [
             "-O3",
-            "-ffp-model=strict",
             "-fprotect-parens",
         ],
     }),
@@ -179,11 +177,9 @@ cc_library(
     copts = select({
         "darwin_arm64_cpu": [
             "-O3",
-            "-ffp-model=strict",
         ],
         "//conditions:default": [
             "-O3",
-            "-ffp-model=strict",
             "-fprotect-parens",
         ],
     }),
@@ -212,7 +208,6 @@ cc_library(
     ],
     copts = [
         "-O3",
-        "-ffp-model=strict",
     ],
     deps = [
         ":riscv_g",
@@ -253,11 +248,9 @@ cc_library(
     copts = select({
         "darwin_arm64_cpu": [
             "-O3",
-            "-ffp-model=strict",
         ],
         "//conditions:default": [
             "-O3",
-            "-ffp-model=strict",
             "-fprotect-parens",
         ],
     }),
@@ -297,7 +290,6 @@ cc_library(
     ],
     copts = [
         "-O3",
-        "-ffp-model=strict",
     ],
     deps = [
         ":riscv_g",
@@ -319,7 +311,6 @@ cc_library(
     ],
     copts = [
         "-O3",
-        "-ffp-model=strict",
     ],
     deps = [
         ":riscv_state",
@@ -340,7 +331,6 @@ cc_library(
         "riscv_zfh_instructions.h",
     ],
     copts = [
-        "-ffp-model=strict",
         "-O3",
     ],
     deps = [
